File: 0206-reverse-linked-list/0206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void rev(ListNode*& t, ListNode*& p, ListNode*& f) {
        if(t == NULL) return;
        f = t->next;
        t->next = p;
        p = t;
        t = f;
        rev(t, p, f);
    }

    ListNode* reverseList(ListNode* head) {
        ListNode* temp = head;
        ListNode* prev = NULL;
        ListNode* front = NULL;
        rev(temp, prev, front);
        return prev;
    }
};

File: 0206-reverse-linked-list/0206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode rev(ListNode t, ListNode p) {
        if(t == null) return p;
        ListNode f = t.next;
        t.next = p;
        p = t;
        t = f;
        return rev(t, p);
    }

    public ListNode reverseList(ListNode head) {
        return rev(head, null);
    }
}